transports/quic/src/transport.rs

```diff
@@ -256,110 +258,119 @@ impl<P: Provider> Transport for GenTransport<P> {
-        let peer_id = peer_id.ok_or(TransportError::MultiaddrNotSupported(addr.clone()))?;
-
-        let socket = self
-            .eligible_listener(&socket_addr)
-            .ok_or(TransportError::Other(
-                Error::NoActiveListenerForDialAsListener,
-            ))?
-            .try_clone_socket()
-            .map_err(Self::Error::from)?;
-
-        tracing::debug!("Preparing for hole-punch from {addr}");
-
-        let hole_puncher = hole_puncher::<P>(socket, socket_addr, self.handshake_timeout);
-
-        let (sender, receiver) = oneshot::channel();
-
-        match self.hole_punch_attempts.entry(socket_addr) {
-            Entry::Occupied(mut sender_entry) => {
-                // Stale senders, i.e. from failed hole punches are not removed.
-                // Thus, we can just overwrite a stale sender.
-                if !sender_entry.get().is_canceled() {
-                    return Err(TransportError::Other(Error::HolePunchInProgress(
-                        socket_addr,
-                    )));
+
+        match (dial_opts.role, dial_opts.port_use) {
+            (Endpoint::Dialer, _) | (Endpoint::Listener, PortUse::New) => {
+                let endpoint = match self.eligible_listener(&socket_addr) {
+                    None => {
+                        // No listener. Get or create an explicit dialer.
+                        let socket_family = socket_addr.ip().into();
+                        let dialer = match self.dialer.entry(socket_family) {
+                            Entry::Occupied(occupied) => occupied.get().clone(),
+                            Entry::Vacant(vacant) => {
+                                if let Some(waker) = self.waker.take() {
+                                    waker.wake();
+                                }
+                                let listen_socket_addr = match socket_family {
+                                    SocketFamily::Ipv4 => {
+                                        SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 0)
+                                    }
+                                    SocketFamily::Ipv6 => {
+                                        SocketAddr::new(Ipv6Addr::UNSPECIFIED.into(), 0)
+                                    }
+                                };
+                                let socket = UdpSocket::bind(listen_socket_addr)
+                                    .map_err(Self::Error::from)?;
+                                let endpoint_config = self.quinn_config.endpoint_config.clone();
+                                let endpoint = Self::new_endpoint(endpoint_config, None, socket)?;
+
+                                vacant.insert(endpoint.clone());
+                                endpoint
+                            }
+                        };
+                        dialer
+                    }
+                    Some(listener) => listener.endpoint.clone(),
```

Comment:
Isn't this the wrong way around here? We shouldn't be using an existing listener when we get `PortUse::New`, right?

Reply:
yes you are right